primitives/src/crh/poseidon/batched_crh.rs

```diff
@@ -1,19 +1,26 @@
 extern crate rand;
 extern crate rayon;
 
-use algebra::{PrimeField, MulShort};
+use algebra::PrimeField;
 use std::marker::PhantomData;
 use crate::crh::BatchFieldBasedHash;
-use crate::{Error, PoseidonParameters, matrix_mix_short, PoseidonHash};
+use crate::{Error, PoseidonParameters, PoseidonHash, PoseidonBatchSBox};
 
-pub struct PoseidonBatchHash<F: PrimeField, P: PoseidonParameters<Fr = F>>{
+pub struct PoseidonBatchHash<F: PrimeField, P: PoseidonParameters<Fr = F>, SB: PoseidonBatchSBox<P>>
+{
     _field:      PhantomData<F>,
     _parameters: PhantomData<P>,
+    _sbox:       PhantomData<SB>,
 }
 
-impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr=F>> PoseidonBatchHash<F, P> {
+impl<F, P, SB> PoseidonBatchHash<F, P, SB>
+    where
+        F: PrimeField,
+        P: PoseidonParameters<Fr = F>,
+        SB: PoseidonBatchSBox<P>,
+{
 
-    fn poseidon_full_round(vec_state: &mut [Vec<P::Fr>], round_cst_idx: &mut usize) {
+    fn poseidon_full_round(vec_state: &mut [Vec<P::Fr>], round_cst_idx: &mut usize, last: bool) {
```

Comment:
The name of this function conflicts the usual names in the context of Poseidon. 
Every round consists of an Add_constants layer, an (partial/full) S-Box layer, and the Matrix-Mix layer. 
I'd separate the adding of the constants from that function, and rebrand it into `fn sbox_layer_full()`. 
The same for the `poseidon_partial_round()` below.

Reply:
Besides that, we will need no boolean to distinguish normal full rounds from the last. 
But I'd expect this to be part of a future PR.

Reply:
Done